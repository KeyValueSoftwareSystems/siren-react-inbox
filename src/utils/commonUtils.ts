import type { RefObject } from "react";

import type {
  ActionResponse,
  NotificationDataType,
  NotificationsApiResponse,
} from "@sirenapp/js-sdk/dist/esm/types";

import {
  defaultBadgeStyle,
  eventTypes,
  LogLevel,
  ThemeMode,
} from "./constants";
import { default as DefaultStyle } from "./defaultStyles";
import { default as DefaultTheme } from "./defaultTheme";
import type {
  CustomStyle,
  DimensionValue,
  SirenStyleProps,
  ThemeProps,
} from "../types";

type FetchParams = {
  size: number;
  start?: string;
  end?: string;
  sort?: "createdAt" | "updatedAt";
};

export const generateElapsedTimeText = (timeString: string) => {
  const currentTime = new Date().getTime();
  const targetTime = new Date(timeString).getTime();
  const millisecondsDiff = currentTime - targetTime;

  const seconds = Math.floor(millisecondsDiff / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);
  const months = Math.floor(days / 30);
  const years = Math.floor(days / 365);

  if (millisecondsDiff < 60000) return "Just now";
  else if (minutes < 60)
    return minutes === 1 ? "1 minute ago" : `${minutes} minutes ago`;
  else if (hours < 24) return hours === 1 ? "1 hour ago" : `${hours} hours ago`;
  else if (days < 30) return days === 1 ? "1 day ago" : `${days} days ago`;
  else if (months < 12)
    return months === 1 ? "1 month ago" : `${months} months ago`;
  else return years === 1 ? "1 year ago" : `${years} years ago`;
};

export const updateNotifications = (
  eventData: {
    id?: string;
    action: string;
    newNotifications?: NotificationDataType[];
    unreadCount?: number;
  },
  notifications: NotificationDataType[]
): NotificationDataType[] => {
  let updatedNotifications: NotificationDataType[] = [];

  switch (eventData.action) {
    case eventTypes.MARK_ITEM_AS_READ:
      updatedNotifications = notifications.map((item) =>
        item.id === eventData.id ? { ...item, isRead: true } : item
      );
      break;
    case eventTypes.MARK_ALL_AS_READ:
      updatedNotifications = notifications.map((item) => ({
        ...item,
        isRead: true,
      }));
      break;
    case eventTypes.DELETE_ITEM:
      updatedNotifications = notifications.filter(
        (item) => item.id !== eventData.id
      );
      break;
    case eventTypes.DELETE_ALL_ITEM:
      updatedNotifications = [];
      break;
    case eventTypes.NEW_NOTIFICATIONS: {
      const newNotifications: NotificationDataType[] =
        eventData?.newNotifications || [];

      updatedNotifications = [...newNotifications, ...notifications];
      break;
    }
    case eventTypes.RESET_NOTIFICATIONS: {
      updatedNotifications = [];
      break;
    }
    default:
      break;
  }

  return updatedNotifications;
};

export const renderAsImage = (icon: JSX.Element | string): boolean =>
  typeof icon === "undefined" || typeof icon === "string";

export const logger = {
  log: async (level: LogLevel.INFO | LogLevel.ERROR, message: string) => {
    const timestamp = new Date().toISOString();
    const levelString = LogLevel[level].toUpperCase();

    console.log(`[${timestamp}] [${levelString}] ${message}`);
  },
  error: function (error: string) {
    this.log(LogLevel.ERROR, error);
  },
  info: function (message: string) {
    this.log(LogLevel.INFO, message);
  },
};

export const applyTheme = (
  theme: ThemeProps = {},
  mode: ThemeMode = ThemeMode.DARK,
  customStyle: CustomStyle = {}
): SirenStyleProps => {
  const windowBorderRadius =
    customStyle.window?.borderRadius || DefaultStyle.window.borderRadius;

  return {
    container: {
      maxWidth: customStyle.window?.width || "100",
    },
    windowShadow: {
      boxShadow: `${
        theme.colors?.windowShadowColor || DefaultTheme[mode].window.shadowColor
      } 0px 8px 24px`,
    },
    windowTopBorder: {
      borderTopLeftRadius: windowBorderRadius,
      borderTopRightRadius: windowBorderRadius,
    },
    windowBottomBorder: {
      borderBottomLeftRadius: windowBorderRadius,
      borderBottomRightRadius: windowBorderRadius,
    },
    contentContainer: {
      backgroundColor:
        theme.windowContainer?.background ||
        theme.colors?.neutralColor ||
        DefaultTheme[mode].windowContainer.background,
      padding:
        customStyle.windowContainer?.padding ||
        DefaultStyle.windowContainer.padding,
      borderRadius: `0 0 ${
        customStyle.window?.borderRadius || DefaultStyle.window.borderRadius
      } ${
        customStyle.window?.borderRadius || DefaultStyle.window.borderRadius
      }`,
    },
    body: {
      overflow: "auto",
      height: customStyle.windowContainer?.contentHeight || "700px",
    },
    headerContainer: {
      alignItems: "center",
      display: "flex",
      margin: 0,
      lineHeight: "28px",
      backgroundColor:
        theme.windowHeader?.background ||
        theme.colors?.neutralColor ||
        DefaultTheme[mode].windowHeader.background,
      borderBottom: `${customStyle.windowHeader?.borderWidth || DefaultStyle.windowHeader.borderWidth} solid`,
      borderColor: theme.colors?.borderColor || DefaultTheme[mode].colors.borderColor,
      height:
        customStyle.windowHeader?.height || DefaultStyle.windowHeader.height,
    },
    headerTitle: {
      color:
        theme.windowHeader?.titleColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].windowHeader.titleColor,
      fontSize:
        customStyle.windowHeader?.titleSize ||
        DefaultStyle.windowHeader.titleSize,
      fontWeight:
        customStyle.windowHeader?.titleFontWeight ||
        DefaultStyle.windowHeader.titleFontWeight,
      paddingLeft:
        customStyle.windowHeader?.titlePadding ||
        DefaultStyle.windowHeader.titlePadding,
    },
    headerAction: {
      color:
        theme.windowHeader?.headerActionColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].windowHeader.headerActionColor,
    },
    defaultCardContainer: {
      backgroundColor:
        theme.notificationCard?.background ||
        DefaultTheme[mode].notificationCard.background,
      padding:
        customStyle.notificationCard?.padding ||
        DefaultStyle.notificationCard.padding,
      borderBottom: `${
        customStyle.notificationCard?.borderWidth ||
        DefaultStyle.notificationCard.borderWidth
      }px solid`,
      borderColor:
        theme.notificationCard?.borderColor ||
        theme.colors?.borderColor ||
        DefaultTheme[mode].notificationCard.borderColor,
    },
    cardIconRound: {
      width:
        customStyle.notificationCard?.avatarSize ||
        DefaultStyle.notificationCard.avatarSize,
      height:
        customStyle.notificationCard?.avatarSize ||
        DefaultStyle.notificationCard.avatarSize,
      borderRadius:
        (customStyle.notificationCard?.avatarSize ||
          DefaultStyle.notificationCard.avatarSize) / 2,
      overflow: "hidden",
      backgroundColor:
        theme.colors?.borderColor || DefaultTheme[mode].colors.borderColor,
    },
    cardTitle: {
      color:
        theme.notificationCard?.titleColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].notificationCard.titleColor,
      fontSize:
        customStyle.notificationCard?.titleSize ||
        DefaultStyle.notificationCard.titleSize,
      fontWeight:
        customStyle.notificationCard?.titleFontWeight ||
        DefaultStyle.notificationCard.titleFontWeight,
    },
    cardSubTitle: {
      color:
        theme.notificationCard?.subTitleColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].notificationCard.subTitleColor,
      fontSize:
        customStyle.notificationCard?.subTitleSize || DefaultStyle.notificationCard.subTitleSize,
      fontWeight:
        customStyle.notificationCard?.subTitleFontWeight ||
        DefaultStyle.notificationCard.subTitleFontWeight,
    },
    activeCardMarker: {
      backgroundColor:
        theme.colors?.highlightedCardColor ||
        DefaultTheme[mode].colors?.highlightedCardColor,
      border:
        theme.colors?.primaryColor || DefaultTheme[mode].colors?.primaryColor,
    },
    cardDescription: {
      color:
        theme.notificationCard?.descriptionColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].notificationCard.descriptionColor,
      fontSize:
        customStyle.notificationCard?.descriptionSize ||
        DefaultStyle.notificationCard.descriptionSize,
      fontWeight:
      customStyle.notificationCard?.descriptionFontWeight ||
      DefaultStyle.notificationCard.descriptionFontWeight,
    },
    dateStyle: {
      color: theme.colors?.dateColor || DefaultTheme[mode].colors.dateColor,
      fontSize:
        customStyle.notificationCard?.dateSize ||
        DefaultStyle.notificationCard.dateSize,
      lineHeight: "16px",
    },
    emptyText: {
      color: theme.colors?.textColor || DefaultTheme[mode].colors.textColor,
    },
    errorText: {
      color: theme.colors?.textColor || DefaultTheme[mode].colors.textColor,
    },
    clearIcon: {
      color: theme.colors?.clearAllIcon || DefaultTheme[mode].clearIcon.color,
      size: customStyle.clearAllIcon?.size || DefaultStyle.clearAllIcon.size,
    },
    timerIcon: {
      color: theme.colors?.timerIcon || DefaultTheme[mode].timerIcon.color,
      size: customStyle.timerIcon?.size || DefaultStyle.timerIcon.size,
    },
    notificationIcon: {
      size:
        customStyle?.notificationIcon?.size ||
        DefaultStyle?.notificationIcon?.size,
    },
    deleteIcon: {
      color: theme.colors?.deleteIcon || DefaultTheme[mode].colors.deleteIcon,
      size: customStyle.deleteIcon?.size || DefaultStyle.deleteIcon.size,
    },
    loadMoreButton: {
      color:
        theme?.loadMoreButton?.color ||
        theme.colors?.primaryColor ||
        DefaultTheme[mode]?.loadMoreButton?.color,
      backgroundColor:
        theme?.loadMoreButton?.background ||
        DefaultTheme[mode]?.loadMoreButton?.background,
      fontSize:
        customStyle?.loadMoreButton?.fontSize ||
        DefaultStyle?.loadMoreButton?.fontSize,
      fontWeight:
        customStyle.loadMoreButton?.fontWeight ||
        DefaultStyle.loadMoreButton.fontWeight,
    },
    loader: {
      backgroundImage: DefaultTheme[mode].loader.backgroundImage,
      borderColor:
        theme.colors?.primaryColor || DefaultTheme[mode].colors.primaryColor,
    },
    badgeStyle: {
      borderRadius: customStyle.badgeStyle?.size || defaultBadgeStyle.size,
      minWidth: customStyle.badgeStyle?.size || defaultBadgeStyle.size,
      height: customStyle.badgeStyle?.size || defaultBadgeStyle.size,
      backgroundColor: theme.badgeStyle?.color || defaultBadgeStyle.color,
      top:  customStyle?.badgeStyle?.top ? `${customStyle.badgeStyle.top}px` : defaultBadgeStyle.top,
      right:  customStyle?.badgeStyle?.right ? `${customStyle.badgeStyle.right}px` : defaultBadgeStyle.right,    
    },
    badgeTextStyle: {
      color: theme.badgeStyle?.textColor || defaultBadgeStyle.textColor,
      fontSize: customStyle.badgeStyle?.textSize || defaultBadgeStyle.textSize,
    },
    infiniteLoader: {
      border: `3px solid ${
        theme.colors?.infiniteLoader ||
        theme.colors?.primaryColor ||
        DefaultTheme[mode].colors.primaryColor
      }`,
    },
  };
};

export const isEmptyArray = (array: Array<NotificationDataType> = []) => {
  return array && array?.length === 0;
};
export const mergeArrays = (
  array1: Array<NotificationDataType> = [],
  array2: Array<NotificationDataType> = []
) => {
  if (array1 && array2) return [...array1, ...array2];
  if (array1) return array1;

  return array2;
};

export const isValidResponse = (
  response: NotificationsApiResponse | ActionResponse
): boolean => !!response?.data;

export const filterDataProperty = (
  response: NotificationsApiResponse
): NotificationDataType[] | null => {
  if (!response.data) return null;

  return response.data;
};

export const generateFilterParams = (
  data: NotificationDataType[],
  fromStart: boolean,
  itemsPerPage: number
): FetchParams => {
  let params: FetchParams = { size: itemsPerPage, sort: "createdAt" };

  if (data.length > 0)
    if (fromStart) params = { ...params, start: data[0].createdAt };
    else params = { ...params, end: data[data.length - 1].createdAt };

  return params;
};

export const calculateModalPosition = (
  iconRef: RefObject<HTMLDivElement>,
  window: Window,
  containerWidth: DimensionValue
) => {
  if (iconRef.current) {
    const iconRect = iconRef.current.getBoundingClientRect();
    const screenWidth = window.innerWidth;
    const spaceRight = screenWidth - iconRect.x;
    const spaceLeft = iconRect.x;
    let modalWidth = calculateModalWidth(containerWidth);

    const centerPosition =
      Math.min(spaceLeft, spaceRight) + Math.abs(spaceLeft - spaceRight) / 2;

    if (window.innerWidth <= modalWidth) modalWidth = window.innerWidth - 40;

    if (spaceRight > modalWidth) {
      return {
        left: `0px`,
      };
    } else if (spaceLeft > modalWidth) {
      const rightPosition = spaceRight < modalWidth ? "30px" : null;

      return {
        ...(rightPosition && { right: rightPosition }),
      };
    } else if (
      spaceLeft < modalWidth &&
      spaceRight < modalWidth &&
      spaceLeft > spaceRight
    ) {
      return { right: "30px" };
    } else {
      return { left: `-${centerPosition - 40}px` };
    }
  }
};

export const calculateModalWidth = (containerWidth: DimensionValue): number => {
  let modalWidth = 500;

  if (typeof containerWidth === "string")
    modalWidth = parseInt(containerWidth.slice(0, -2)) + 40;
  else if (typeof containerWidth === "number") modalWidth = containerWidth + 40;

  return modalWidth;
};

export const hexToRgba = (hex: string, alpha: number) => {
  const [r, g, b] = hex.match(/\w\w/g)?.map((x) => parseInt(x, 16)) ?? [];

  return `rgba(${r},${g},${b},${alpha})`;
};

export const debounce = <F extends (...args: unknown[]) => void>(
  func: F,
  delay: number
) => {
  let timerId: ReturnType<typeof setTimeout>;

  return (...args: Parameters<F>): void => {
    clearTimeout(timerId);
    timerId = setTimeout(() => {
      func(...args);
    }, delay);
  };
};
