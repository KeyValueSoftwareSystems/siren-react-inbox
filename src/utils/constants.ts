export const COLORS = {
  light: {
    primaryColor: "#F56630",
    highlightedCardColor: "#FFECE5",
    textColor: "#344054",
    neutralColor: "#FFFFFF",
    borderColor: "#D0D5DD",
    dateColor: "#34405499",
    deleteIcon: "#34405499",
    timerIcon: "#667185",
    clearAllIcon: "#667185",
    windowShadowColor: "#D0D5DD",
    iconColor: "#F7F9FC",
    notificationIcon: "#232326",
  },
  dark: {
    primaryColor: "#F56630",
    highlightedCardColor: "#2E2D30",
    textColor: "#FFFFFF",
    neutralColor: "#232326",
    borderColor: "#344054",
    dateColor: "#98A2B3",
    deleteIcon: "#D0D5DD",
    timerIcon: "#98A2B3",
    clearAllIcon: "#D0D5DD",
    windowShadowColor: "#232326",
    iconColor: "#38383D",
    notificationIcon: "#FFFFFF",
  },
};

export enum ThemeMode {
  DARK = "dark",
  LIGHT = "light",
}

export enum LogLevel {
  INFO,
  ERROR,
}

export enum BadgeType {
  NONE = "none",
  DOT = "dot",
  DEFAULT = "default",
}

export const levelLogFns = {
  [LogLevel.INFO]: console.log,
  [LogLevel.ERROR]: console.error,
};

export const defaultBadgeStyle = {
  size: 15,
  color: "#FF0000",
  textColor: "#FFFFFF",
  textSize: 10,
  linHeight: "14px",
  top: "inherit",
  left: "inherit",
};

export enum eventTypes {
  MARK_ITEM_AS_VIEWED = "MARK_ITEM_AS_VIEWED",
  MARK_ITEM_AS_READ = "MARK_ITEM_AS_READ",
  MARK_ALL_AS_READ = "MARK_ALL_AS_READ",
  DELETE_ITEM = "DELETE_ITEM",
  DELETE_ALL_ITEM = "DELETE_ALL_ITEM",
  NEW_NOTIFICATIONS = "NEW_NOTIFICATIONS",
  UPDATE_NOTIFICATIONS_COUNT = "UPDATE_NOTIFICATIONS_COUNT",
  RESET_NOTIFICATIONS = "RESET_NOTIFICATIONS",
  RESET_NOTIFICATIONS_COUNT = "RESET_NOTIFICATIONS_COUNT",
}

export enum events {
  NOTIFICATION_LIST_EVENT = "NOTIFICATION_LIST_EVENT",
  NOTIFICATION_COUNT_EVENT = "NOTIFICATION_COUNT_EVENT",
}

export enum updateNotificationsTypes {
  MARK_ITEM_AS_VIEWED = "MARK_ITEM_AS_VIEWED",
  MARK_ITEM_AS_READ = "MARK_ITEM_AS_READ",
  MARK_ALL_AS_READ = "MARK_ALL_AS_READ",
  DELETE_ITEM = "DELETE_ITEM",
  DELETE_ALL_ITEM = "DELETE_ALL_ITEM",
}

export enum sirenReducerTypes {
  SET_NOTIFICATIONS = "SET_NOTIFICATIONS",
  NEW_NOTIFICATIONS = "NEW_NOTIFICATIONS",
  SET_UN_VIEWED_NOTIFICATION_COUNT = "SET_UN_VIEWED_NOTIFICATION_COUNT",
  SET_SIREN_CORE = "SET_SIREN_CORE",
}

export const LIST_EMPTY_TEXT = "No new notifications";
export const LIST_EMPTY_SUB_TEXT = "Check back later for updates and alerts.";
export const ERROR_TEXT = "Oops! Something went wrong";
export const ERROR_SUB_TEXT =
  "Could not load the notifications. Please refresh the page.";
export const DEFAULT_WINDOW_TITLE = "Notifications";
export const RETRY_BUTTON_LABEL = "Retry";
export const CLEAR_ALL_LABEL = "Clear All";
export const IN_APP_RECIPIENT_UNAUTHENTICATED = 'IN_APP_RECIPIENT_UNAUTHENTICATED';
export const TOKEN_VERIFICATION_PENDING ='TOKEN_VERIFICATION_PENDING';
export const MAXIMUM_RETRY_COUNT = 3;

export const errorMap = {
  SIREN_OBJECT_NOT_FOUND: {
    Type: "ERROR",
    Code: "SIREN_OBJECT_NOT_FOUND",
    Message: "Siren Object Not found",
  },
  MISSING_PARAMETER: {
    Type: "ERROR",
    Code: "MISSING_PARAMETER",
    Message: "Missing Parameter",
  },
};

export enum VerificationStatus {
  PENDING = "PENDING",
  SUCCESS = "SUCCESS",
  FAILED = "FAILED"
}