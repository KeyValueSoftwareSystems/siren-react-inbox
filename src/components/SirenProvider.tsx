import React, { createContext, useContext, useEffect, useState } from "react";

import { Siren } from "@sirenapp/js-sdk";
import type {
  InitConfigType,
  NotificationsApiResponse,
  SirenErrorType,
  UnviewedCountApiResponse,
} from "@sirenapp/js-sdk/dist/esm/types";
import PubSub from "pubsub-js";

import type { SirenProviderConfigProps } from "../types";
import { logger } from "../utils/commonUtils";
import {
  AUTHENTICATION_FAILED,
  events,
  EventType,
  eventTypes,
  IN_APP_RECIPIENT_UNAUTHENTICATED,
  MAXIMUM_RETRY_COUNT,
  VerificationStatus,
} from "../utils/constants";

type SirenContextProp = {
  siren: Siren | null;
  verificationStatus: VerificationStatus;
};

interface SirenProvider {
  config: SirenProviderConfigProps;
  children: React.ReactNode;
}

export const SirenContext = createContext<SirenContextProp>({
  siren: null,
  verificationStatus: VerificationStatus.PENDING,
});

/**
 * Use `useSirenContext` hook to access Siren notifications context within your component.
 *
 * @example
 * const {
 *   siren,
 *   verificationStatus
 * } = useSirenContext();
 *
 * @returns {SirenContextProp} The Siren notifications context.
 */
export const useSirenContext = (): SirenContextProp => useContext(SirenContext);

/**
 * Provides a React context for Siren notifications, making Siren SDK functionality
 * available throughout your React application.
 *
 * `SirenProvider` initializes the Siren SDK with given configuration and manages the state for siren and verificationStatus.
 *
 * @component
 * @example
 * const config = {
 *   userToken: "user_token_here",
 *   recipientId: "recipient_id_here"
 * };
 *
 * <SirenProvider config={config}>
 *   <YourComponent />
 * </SirenProvider>
 *
 * @param {Object} props - Props for configuring the SirenProvider.
 * @param {SirenProviderConfig} props.config - Configuration for initializing the Siren SDK.
 * @param {React.ReactNode} props.children - Child components that will have access to the Siren context.
 */
const SirenProvider: React.FC<SirenProvider> = ({ config, children }) => {
  const [siren, setSiren] = useState<Siren | null>(null);
  const [verificationStatus, setVerificationStatus] =
    useState<VerificationStatus>(VerificationStatus.PENDING);
  let retryCount = 0;

  useEffect(() => {
    if (config?.recipientId && config?.userToken) {
      stopRealTimeFetch();
      sendResetDataEvents();
      initialize();
    }
    else {
      setVerificationStatus(VerificationStatus.FAILED);
    }
    if (retryCount > MAXIMUM_RETRY_COUNT) stopRealTimeFetch();
  }, [config]);

  const stopRealTimeFetch = (): void => {
    siren?.stopRealTimeFetch(EventType.NOTIFICATION);
    siren?.stopRealTimeFetch(EventType.UNVIEWED_COUNT);
  };

  const sendResetDataEvents = () => {
    const updateCountPayload = {
      action: eventTypes.RESET_NOTIFICATIONS_COUNT,
    };
    const updateNotificationPayload = {
      action: eventTypes.RESET_NOTIFICATIONS,
    };

    PubSub.publish(
      events.NOTIFICATION_COUNT_EVENT,
      JSON.stringify(updateCountPayload)
    );
    PubSub.publish(
      events.NOTIFICATION_LIST_EVENT,
      JSON.stringify(updateNotificationPayload)
    );
  };

  const onEventReceive = (
    response: NotificationsApiResponse | UnviewedCountApiResponse,
    eventType: EventType
  ): void => {
    if (eventType === EventType.NOTIFICATION) {
      const newNotifications =
        (response as NotificationsApiResponse)?.data || [];

      logger.info(`new notifications : ${JSON.stringify(newNotifications)}`);
      PubSub.publish(
        events.NOTIFICATION_LIST_EVENT,
        JSON.stringify({
          newNotifications,
          action: eventTypes.NEW_NOTIFICATIONS,
        })
      );
    }

    if (eventType === EventType.UNVIEWED_COUNT) {
      const totalUnviewed =
        (response as UnviewedCountApiResponse)?.data?.totalUnviewed || 0;

      logger.info(`unviewed notification count : ${totalUnviewed}`);
      PubSub.publish(
        events.NOTIFICATION_COUNT_EVENT,
        JSON.stringify({
          unviewedCount: totalUnviewed,
          action: eventTypes.UPDATE_NOTIFICATIONS_COUNT,
        })
      );
    }
  };

  const onStatusChange = (status: VerificationStatus) => {
    setVerificationStatus(status);
  };

  const actionCallbacks = { onEventReceive, onStatusChange };

  const getDataParams = () => {
    return {
      token: config.userToken,
      recipientId: config.recipientId,
      onError: retryVerification,
      actionCallbacks: actionCallbacks,
    };
  };

  const retryVerification = (error: SirenErrorType) => {
    const shouldRetry = (error.Code === AUTHENTICATION_FAILED || error.Code === IN_APP_RECIPIENT_UNAUTHENTICATED) &&
      retryCount < MAXIMUM_RETRY_COUNT && verificationStatus === VerificationStatus.FAILED

    if (shouldRetry)
      setTimeout(() => {
        initialize();
        retryCount++;
      }, 5000);
  };

  // Function to initialize the Siren SDK and fetch notifications
  const initialize = (): void => {
    const dataParams: InitConfigType = getDataParams();
    const siren = new Siren(dataParams);

    setVerificationStatus(VerificationStatus.PENDING);
    setSiren(siren);
  };

  return (
    <SirenContext.Provider
      value={{
        siren,
        verificationStatus,
      }}
    >
      {children}
    </SirenContext.Provider>
  );
};

export default SirenProvider;
