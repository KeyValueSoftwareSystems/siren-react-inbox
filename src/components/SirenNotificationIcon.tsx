import React, { type FC, useCallback, useEffect, useState } from "react";

import BellIcon from "./BellIcon";
import { useSirenContext } from "./SirenProvider";
import "../styles/sirenNotificationIcon.css";
import type { SirenNotificationButtonProps } from "../types";
import { Constants } from "../utils";
import { BadgeType, EventType, MAX_UNVIEWED_COUNT_SHOWN } from "../utils/constants";

const { eventTypes, events } = Constants;

/**
 * SirenNotificationIcon Component
 * @component
 * @param {Object} props - Props for the SirenNotificationIcon component
 * @param {ReactNode} [props.notificationIcon] - The icon for the notification
 * @param {BadgeType} [props.badgeType=DEFAULT] - The type of badge to display
 * @param {Object} [props.styles] - Custom styles for the component
 * @param {Function} props.onIconClick - Click event handler for the icon
 * @param {boolean} [props.hideBadge] - Flag indicating if the badge should be hidden
 * @returns {JSX.Element} - SirenNotificationIcon component JSX
 */

const SirenNotificationIcon: FC<SirenNotificationButtonProps> = ({
  notificationIcon,
  styles,
  onIconClick,
  darkMode,
  hideBadge,
  isModalOpen,
}) => {
  const { siren } = useSirenContext();

  const [unviewedCount, seUnviewedCount] = useState<number>(0);
  const badgeType:BadgeType = isModalOpen ? BadgeType.NONE : BadgeType.DEFAULT;
  
  const notificationCountSubscriber = async (
    type: string,
    dataString: string
  ) => {
    const data = await JSON.parse(dataString);

    if (data.action === eventTypes.UPDATE_NOTIFICATIONS_COUNT)
      seUnviewedCount(data.unviewedCount);
  };

  useEffect(() => {
    if(!hideBadge) {
      PubSub.subscribe(
        events.NOTIFICATION_COUNT_EVENT,
        notificationCountSubscriber
      );

      return () => {
        cleanUp();
      };
    }
  }, []);

  useEffect(() => {
    // Check to avoid calling getUnViewedCount when the badge is hidden and the modal is open, and when either the siren object or hideBadge state changes.
    if(!hideBadge && !isModalOpen) 
      getUnViewedCount();
    
  }, [siren, hideBadge]);

  useEffect(() => {
    if(!hideBadge)
      startRealTimeDataFetch();
  }, [hideBadge]);

  const cleanUp = () => {
    siren?.stopRealTimeFetch(EventType.UNVIEWED_COUNT);
  };

  const startRealTimeDataFetch = (): void => {
    cleanUp();
    siren?.startRealTimeFetch({eventType: EventType.UNVIEWED_COUNT});
  };

  const getUnViewedCount = async (): Promise<void> => {
    if (!siren) return;
    try {
      const response = await siren.fetchUnviewedNotificationsCount();

      startRealTimeDataFetch();
      if (response && response.error) return;
      if (response?.data) seUnviewedCount(response?.data?.unviewedCount || 0);
    } catch (er) {
      //  handle error if needed
    }
  };

  const renderCount = useCallback(
    () => (unviewedCount > MAX_UNVIEWED_COUNT_SHOWN ? `${MAX_UNVIEWED_COUNT_SHOWN}+` : unviewedCount),
    [unviewedCount]
  );

  const renderBadge = () => {
    switch (badgeType) {
      case BadgeType.DEFAULT: {
        return (
          unviewedCount > 0 && (
            <div
              style={styles.badgeStyle}
              className="siren-sdk-notificationIcon-badge-container"
            >
              <div
                style={styles.badgeTextStyle}
                data-testid="notification-default-badge"
              >
                {renderCount()}
              </div>
            </div>
          )
        );
      }

      default:
        return null;
    }
  };

  return (
    <button
      onClick={onIconClick}
      className="siren-sdk-notificationIcon-container"
      data-testid="notification-icon"
      aria-label="siren-notification-icon"
    >
      {notificationIcon || (
        <BellIcon
          size={styles.notificationIcon.size}
          color={
            darkMode
              ? Constants.COLORS.dark.notificationIcon
              : Constants.COLORS.light.notificationIcon
          }
        />
      )}
      {!hideBadge && renderBadge()}
    </button>
  );
};

export default SirenNotificationIcon;
