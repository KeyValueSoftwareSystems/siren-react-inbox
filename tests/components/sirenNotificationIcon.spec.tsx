import React from "react";

import { fireEvent, render } from "@testing-library/react";

import SirenNotificationIcon from "../../src/components/SirenNotificationIcon";
import { SirenContext } from "../../src/components/SirenProvider"; // Assuming SirenProvider exports SirenContext
import { applyTheme } from "../../src/utils/commonUtils";
import { VerificationStatus } from "../../src/utils/constants";

const mockClickFn = jest.fn();

// Mock the CSS file to avoid Jest error
jest.mock("../../src/styles/sirenNotificationIcon.css", () => ({}));

const mockSirenContextValue = {
  siren: null,
  verificationStatus: VerificationStatus.SUCCESS
};

const style = applyTheme();

test("matches snapshot", () => {
  const { asFragment } = render(
    <SirenNotificationIcon
      styles={style}
      notificationIcon={<div />}
      onIconClick={mockClickFn}
      darkMode={false}
      hideBadge={false}
      isModalOpen={false}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});

test("should render notification icon only", () => {
  const component = render(
    <SirenNotificationIcon
      styles={style}
      onIconClick={mockClickFn}
      notificationIcon={<div />}
      darkMode={true}
      hideBadge={false}
      isModalOpen={false}
    />
  );
  const badge = component.queryByTestId("notification-default-badge");

  expect(badge).toBeNull();
});

test("should render the passed notification icon", () => {
  const component = render(
    <SirenNotificationIcon
      styles={style}
      onIconClick={mockClickFn}
      notificationIcon={<div data-testid="custom-notification-icon" />}
      darkMode={true}
      hideBadge={false}
      isModalOpen={false}
    />
  );
  const button = component.getByTestId("custom-notification-icon"); // Assuming count is displayed as button text

  expect(button).toBeTruthy();
});

test("should call onClick handler on button click", () => {
  const component = render(
    <SirenNotificationIcon
      styles={style}
      onIconClick={mockClickFn}
      notificationIcon={<div />}
      isModalOpen={false}
      darkMode={true}
      hideBadge={false}
    />
  );
  const button = component.getByTestId("notification-icon");

  fireEvent.click(button);

  expect(mockClickFn).toHaveBeenCalledTimes(1);
});

it("does not render badge with count if unviewed count is 0", () => {
  const { queryByTestId } = render(
    <SirenNotificationIcon
      styles={style}
      onIconClick={mockClickFn}
      notificationIcon={<div />}
      darkMode={true}
      hideBadge={false}
      isModalOpen={false}
    />
  );
  const badge = queryByTestId("notification-default-badge");

  expect(badge).toBeFalsy();
});



it("calls onClick handler when icon is clicked", () => {
  const onClickMock = jest.fn();
  const { getByTestId } = render(
    <SirenNotificationIcon
      onIconClick={onClickMock}
      styles={style}
      notificationIcon={<div />}
      isModalOpen={false}
      darkMode={true}
      hideBadge={false}
    />
  );
  const icon = getByTestId("notification-icon");

  fireEvent.click(icon);
  expect(onClickMock).toHaveBeenCalledTimes(1);
});

test("should not render badge with count if unviewed count is 0", () => {
  const { queryByTestId } = render(
    <SirenNotificationIcon
      styles={style}
      onIconClick={mockClickFn}
      notificationIcon={<div />}
      darkMode={true}
      hideBadge={false}
      isModalOpen={false}
    />
  );
  const badge = queryByTestId("notification-default-badge");

  expect(badge).toBeFalsy();
});

test("should hide badge if unviewed count is 0 and badge type is default", () => {
  const { queryByTestId } = render(
    <SirenNotificationIcon
      styles={style}
      onIconClick={mockClickFn}
      notificationIcon={<div />}
      darkMode={true}
      hideBadge={false}
      isModalOpen={false}
    />
  );
  const badge = queryByTestId("notification-default-badge");

  expect(badge).toBeNull();
});

test("should hide badge if badge type is set to none", () => {
  const { queryByTestId } = render(
    <SirenNotificationIcon
      styles={style}
      onIconClick={mockClickFn}
      isModalOpen={false}
      notificationIcon={<div />}
      darkMode={true}
      hideBadge={false}
    />
  );
  const badge = queryByTestId("notification-default-badge");

  expect(badge).toBeNull();
});

test("calls PubSub.subscribe with correct arguments", () => {
  render(
    <SirenNotificationIcon
      styles={style}
      notificationIcon={<div />}
      onIconClick={mockClickFn}
      darkMode={true}
      hideBadge={false}
      isModalOpen={false}
    />,
    {
      wrapper: ({ children }) => (
        <SirenContext.Provider value={mockSirenContextValue}>
          {children}
        </SirenContext.Provider>
      ),
    }
  );
});