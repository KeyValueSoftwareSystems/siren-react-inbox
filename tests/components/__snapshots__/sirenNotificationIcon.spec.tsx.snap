// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="siren-notification-icon"
    class="siren-sdk-notificationIcon-container"
    data-testid="notification-icon"
  >
    <div />
  </button>
</DocumentFragment>
`;
